### PR TITLE
add get_dataset_permissions and set_dataset_permissions functions

### PR DESCRIPTION
**Pull request by FredericBGA**
let the user get and set permissions at the dataset level

### CODE REVIEW

**Comment by FredericBGA** [comment]
ok, I need help for the tests.

first: how can I run them locally? At least for 19.05

then, I don't understand the failure  ([release_19.05](https://travis-ci.org/galaxyproject/bioblend/jobs/575333184)) about the tearDown function. I used to code some tests for Django apps.

I though that tearDown and setUp were called for each test function. So I don't understand here  why tearDown has no library to delete. The setUp should have created one, no?

for some ([at least this one: ](https://travis-ci.org/galaxyproject/bioblend/jobs/575333201) ) the error is:

> The` mandatory parameter \"action\" is missing
For this one I noticed in the code that I must set an action in payload (see my review above)

So there differences between Galaxy versions. How can I handle this? (with `test_util.skip_unless_galaxy('release_xx.xx')`)

thank you.

File: bioblend/galaxy/libraries/__init__.py [hunk 1]
```diff
@@ -602,10 +602,25 @@ def get_library_permissions(self, library_id):
...
+        """
+        Get the permissions for a dataset.
+
+        :type dataset_id: str
+        :param dataset_id: id of the dataset
+
+        :rtype: dict
+        :return: dictionary with all applicable permissions' values
+        """
+        # change module and use datasets API
```
Comment (nsoranzo) [inline bioblend/galaxy/libraries/__init__.py]:
```suggestion
```

File: bioblend/galaxy/libraries/__init__.py [hunk 2]
```diff
@@ -637,3 +652,38 @@ def set_library_permissions(self, library_id, access_in=None,
...
+            payload['access_ids[]'] = access_in
+        if modify_in:
+            payload['modify_ids[]'] = modify_in
+        if manage_in:
+            payload['manage_ids[]'] = manage_in
+        # we need here to define an action
+        payload['action'] = 'set_permissions'
+        # change module and use datasets API
+        self.module = 'libraries/datasets'
+        url = '/'.join([self.gi._make_url(self, dataset_id), 'permissions'])
```
Comment (nsoranzo) [inline bioblend/galaxy/libraries/__init__.py]:
```suggestion
        url = '/'.join([self.gi.url, 'libraries/datasets', dataset_id, 'permissions'])
```

File: bioblend/_tests/TestGalaxyLibraries.py [hunk 3]
```diff
@@ -102,3 +102,13 @@ def test_library_permissions(self):
...
+
+    def test_dataset_permissions(self):
+        current_user = self.gi.users.get_current_user()
+        user_id_list_new = [current_user['id']]
+        library_id = self.library["id"]
+        dataset1 = self.gi.libraries.upload_file_contents(library_id, FOO_DATA)
+        ret = self.gi.libraries.set_dataset_permissions(dataset1[0]['id'], access_in=user_id_list_new, modify_in=user_id_list_new, manage_in=user_id_list_new)
+        self.assertEqual(set(_[1] for _ in ret['access_dataset_roles']), set(user_id_list_new))
+        self.assertEqual(set(_[1] for _ in ret['modify_item_roles']), set(user_id_list_new))
+        self.assertEqual(set(_[1] for _ in ret['manage_dataset_roles']), set(user_id_list_new))
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyLibraries.py]: Can you also test `self.gi.libraries.get_dataset_permissions()` here, please?

File: bioblend/galaxy/libraries/__init__.py [hunk 4]
```diff
@@ -637,3 +652,38 @@ def set_library_permissions(self, library_id, access_in=None,
...
+        if access_in:
+            payload['access_ids[]'] = access_in
+        if modify_in:
+            payload['modify_ids[]'] = modify_in
+        if manage_in:
+            payload['manage_ids[]'] = manage_in
+        # we need here to define an action
+        payload['action'] = 'set_permissions'
+        # change module and use datasets API
+        self.module = 'libraries/datasets'
```
[the remark this answers is not in the document]
Comment (nsoranzo) [reply bioblend/galaxy/libraries/__init__.py]: You've missed this ^

**Comment by FredericBGA** [comment]
thank you for your help @nsoranzo 
I hope it will be helpful.

**Pull request merged**